### PR TITLE
Fastly edge caching: use PURGE HTTP method instead of POST

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR adds a feature flag called `fastly_http_purge` that we can use to enable the `PURGE` HTTP method to [purge an individual URL](https://github.com/fastly/fastly-ruby#efficient-purging).

Right now we're using the deprecated "purge by `POST`", we'd like to switch to the new method.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/12326

## QA Instructions, Screenshots, Recordings

It's not possible to test this locally, I think we should deploy to https://community.benhalpern.com/ which uses Fastly, enable the flag in the `/admin/feature_flags/features` UI and then excersise caching to make sure stuff is purged.

We can use the `Fastly-Debug` header to test it, for example:

```shell
curl -H "Fastly-Debug: 1" -I https://community.benhalpern.com/{path}
```

## Added tests?

- [x] Yes (also: automatic tests for Fastly are moot as they can't be run locally)
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] README
- [x] No documentation needed

### CODE REVIEW

File: lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb
Comment (rhymes) [inline lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb:4]: I think we should get in the habit of adding/removing feature flags programmatically so that we know each Forem is aware of them

File: app/services/edge_cache/bust/fastly.rb
Comment (rhymes) [inline app/services/edge_cache/bust/fastly.rb:23]: using [fastly-ruby gem here](https://github.com/fastly/fastly-ruby#efficient-purging). The [implementation is more or less what we would have written anyway](https://github.com/fastly/fastly-ruby/blob/acae3028a079f9bf0cbea4eaa349a0853441053d/lib/fastly/client.rb#L92-L107) and we get `Thread.current` [for free](https://github.com/fastly/fastly-ruby/blob/acae3028a079f9bf0cbea4eaa349a0853441053d/lib/fastly/client.rb#L111-L113)

Comment (atsmith813) [reply app/services/edge_cache/bust/fastly.rb:23]: Great call! I'm so stoked 🔥

File: app/services/edge_cache/bust/fastly.rb
Comment (atsmith813) [inline app/services/edge_cache/bust/fastly.rb:6]: 🔥

File: app/services/edge_cache/bust/fastly.rb
Comment (atsmith813) [inline app/services/edge_cache/bust/fastly.rb:33]: 🤩

**Review by atsmith813 — APPROVED** [review]

File: app/services/edge_cache/bust/fastly.rb
Comment (citizen428) [reply app/services/edge_cache/bust/fastly.rb:23]: @rhymes Is there a reason why we can't also use the gem in `fastly_post`? It seems odd to me to use `HTTParty` in one method and the gem in the other, but maybe I'm missing something.

File: app/services/edge_cache/bust/fastly.rb
Comment (citizen428) [inline app/services/edge_cache/bust/fastly.rb:30]: Could we use our helper here?

```suggestion
          URL.url(path),
```

File: app/services/edge_cache/bust/fastly.rb
Comment (citizen428) [inline app/services/edge_cache/bust/fastly.rb:31]: Same as above:

```suggestion
          URL.url("#{path}?i=i"),
```

File: lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb
Comment (citizen428) [reply lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb:4]: This may be worth documenting, though we may not have any documentation on feature flags so far.

File: spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb
Comment (citizen428) [inline spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb:12]: We could change (no pun intended) this to use a `change` assertion:

```ruby
expect do
  described_class.new.run
end.to change { FeatureFlag.exist?(:fastly_http_purge) }.from(true).to(false)
```

File: spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb
Comment (citizen428) [inline spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb:24]: Similar to above:

```suggestion
    expect do
      described_class.new.run
    end.not_to change { FeatureFlag.exist?(:fastly_http_purge) }
```

File: app/services/edge_cache/bust/fastly.rb
Comment (rhymes) [reply app/services/edge_cache/bust/fastly.rb:23]: @citizen428 I thought about it but I decided to only change one thing at a time, since `HTTParty.post` works. Also the objective here is to remove that code, so I didn't bother checking if there were any differences :-)

File: lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb
Comment (rhymes) [reply lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb:4]: @citizen428 where should we add it? Most feature flags are fleeting but I guess we can keep an internal list of them.

File: spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb
Comment (rhymes) [reply spec/lib/data_update_scripts/add_fastly_http_purge_feature_flag_spec.rb]: thanks!

File: lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb
Comment (citizen428) [reply lib/data_update_scripts/20210209152726_add_fastly_http_purge_feature_flag.rb:4]: Sorry, I meant documenting that they should be added before being enabled, not the specific flags.

**Review by citizen428 — APPROVED** [review]
As far as I'm concerned this is good to go once it gets out of draft, thanks @rhymes!

**Review by maestromac — APPROVED** [review]
Somewhat tested on Ben Halpern community. LGTM! seeing that it's behind feature flag it is safe to deploy

**Comment by leewynne** [comment]
Hi there! I also use Fastly on TMMAL. Is there anything I need to do before merging this?

**Comment by maestromac** [comment]
@leewynne nope you can safely merge this

**Pull request merged**
